include/boost/gil/image_processing/threshold.hpp

```diff
@@ -0,0 +1,85 @@
+//
+// Copyright 2019 Miral Shah <miralshah2211@gmail.com>
+//
+// Use, modification and distribution are subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//
+#ifndef BOOST_GIL_THRESHOLD_HPP
+#define BOOST_GIL_THRESHOLD_HPP
+
+#include <functional>
+#include <boost/gil/image.hpp>
+
+namespace boost { namespace gil {
+
+
+namespace detail {
+
+template <typename source_channel_t, typename result_channel_t,
+    typename SrcView, typename DstView>
+void threshold_simple_impl(SrcView const& src, DstView const& dst, int threshold,
+    int max, const std::function<result_channel_t(source_channel_t)>& logic)
+{
+    //template argument validation
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    gil_function_requires<MutableImageViewConcept<DstView>>();
+    gil_function_requires<ColorSpacesCompatibleConcept<
+        typename color_space_type<SrcView>::type,
+        typename color_space_type<DstView>::type>>();
+
+    //deciding output channel type and creating functor
+    typedef typename channel_type<SrcView>::type source_channel_t;
+    typedef typename channel_type<DstView>::type result_channel_t;
+
+    //iterate over the image chaecking each pixel value for the threshold
+    for (size_t y = 0; y < src.height(); y++)
+    {
+        typename SrcView::x_iterator src_it = src.row_begin(y);
+        typename DstView::x_iterator dst_it = dst.row_begin(y);
+
+        for (size_t x = 0; x < src.width(); x++)
+        {
+            static_transform(src_it[x], dst_it[x], logic);
+        }
+    }
+}
+
+} //namespace boost::gil::detail
+
+//threshold takes an image view and perform thresholding operation on each channel
+//and returns and image view.
+template <typename SrcView, typename DstView>
+void threshold_binary(SrcView const& src, DstView const& dst, int threshold, int max)
```

Comment:
1. `src` -> `src_???` Should be `src_view`

2. `max` could have a better name.

3. More on `max`

    * Can `max` be derived automatically from channel type?
    * Should it be derived or is it better to let user to specify it?

Reply:
how is `max_value` ?

yes, max can be derived from channel type. We can set max to the maximum value of channel by default and let the user specify if required.

Reply:
what you suggested can be achieved in this way although it makes the code more complex.

```
template <typename SrcView, typename DstView>
void threshold_binary_inverse
(
    SrcView const& src_view,
    DstView const& dst_view,
    int threshold,
    typename channel_type<DstView>::type max=std::numeric_limits<typename channel_type<DstView>::type>::max(),
    bool low=true
)
```

Reply:
Right. Lets stick to explicit `max_value` (it's  better name indeed).

I was thinking of having overloads like this 

```
threshold_binary(SrcView const& src_view, DstView const& dst_view, int threshold);
threshold_binary(SrcView const& src_view, DstView const& dst_view, int threshold, double max_value);
```
where the former calls the latter with internally determined `max_value` based on `channel_type<DstViev>`.

### Type of `max_value`

In case of the function with explicit `max_value`, I think there are two options regarding type of this parameter:
* `double` as tape if enough capacity to cover all possible channels (with `static_cast` internally into proper type for `DstView`)
* `typename channel_type<DstView>::type` a

Reply:
double is indeed enough I have never seen any image with more channel depth(max depth used now is 48bit which is still rare). Although I prefer the second option just to avoid conversions(to avoid `warning` such as `possible data loss`).

Reply:
Yes, good point.